safemode/safemode.go

```diff
@@ -0,0 +1,242 @@
+
+	if disruption.Spec.Network != nil {
+		safemodeNetwork := Network{}
+		safemodeNetwork.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeNetwork)
+	}
+
+	if disruption.Spec.DiskPressure != nil {
+		safemodeDisk := Disk{}
+		safemodeDisk.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeDisk)
+	}
+
+	if disruption.Spec.ContainerFailure != nil {
+		safemodeContainerFailure := ContainerFailure{}
+		safemodeContainerFailure.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeContainerFailure)
+	}
+
+	if disruption.Spec.CPUPressure != nil {
+		safemodeCPU := CPU{}
+		safemodeCPU.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeCPU)
+	}
+
+	if disruption.Spec.DNS != nil {
+		safemodeDNS := DNS{}
+		safemodeDNS.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeDNS)
+	}
+
+	if disruption.Spec.GRPC != nil {
+		safemodeGRPC := GRPC{}
+		safemodeGRPC.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeGRPC)
+	}
+
+	if disruption.Spec.NodeFailure != nil {
+		safemodeNode := Node{}
+		safemodeNode.Init(disruption, k8sClient)
+		safemodeList = append(safemodeList, &safemodeNode)
+	}
+
+	return safemodeList
+}
+
+type Generic struct {
+	dis    v1beta1.Disruption
+	client client.Client
+}
+
+// Init Refer to safemode.Safemode interface for documentation
+func (sm *Generic) Init(disruption v1beta1.Disruption, client client.Client) {
+	sm.dis = disruption
+	sm.client = client
+}
+
+// CreationSafetyNets Refer to safemode.Safemode interface for documentation
+func (sm *Generic) CreationSafetyNets() ([]string, error) {
+	safetyNetResponses := []string{}
+
+	if caught, err := sm.safetyNetCountNotTooLarge(); err != nil {
+		return nil, err
+	} else if caught {
+		safetyNetResponses = append(safetyNetResponses, " The specified count represents a large percentage of targets in either the namespace or the kubernetes cluster")
+	}
+
+	if caught, err := sm.safetyNetSporadicTargets(); err != nil {
+		return nil, err
+	} else if caught {
+		safetyNetResponses = append(safetyNetResponses, " The target environment's size is changing sporadically, this is a sign of instability and applying additional disruption can be dangerous")
+	}
+
+	return safetyNetResponses, nil
+}
+
+// safetyNetSporadicTargets is the safety net regarding sporadic targets
+// in an environment which count is constantly changing, we are looking at an environment where targets are being
+// destroyed and created frequently which is a very bad sign in terms of reliability and a applying further
+// disruptions is probably not the best idea.
+// In this function we run a check against the count of the target environment 3 times. If the count is different each of those
+// 3 times, we assume sporadic behaviour of the target environment and raise a flag.
+func (sm *Generic) safetyNetSporadicTargets() (bool, error) {
```

Comment:
Are those 3 list calls done sequentially (with no pause) and on disruption creation? Because if it is the case, the time to pull the injector image and to effectively inject the disruption, we may list targets before the injection is really done. In addition to this, if we take the example of a pod with a liveness probe, the pod may start failing multiple seconds after the disruption is injected and you will likely miss that.

I'd also not only rely on the targets count but maybe on the targets status changes? A target going from ready to not ready during a disruption (as well as a target being replaced as you check here) is a good indicator of something going wrong. Though I feel those things should be checked during the whole disruption for effects that may take a while to appear. And in this case, relying on events may be a better solution so you can follow changes over time (like we do for dynamic target injection in Nathan's PR).

Reply:
The more I think about this safety net, the more I realize that it should not be a satefy net. The idea of such a mechanism is very good to be honest (I'd love to see the controller telling to someone that the disruption they created somehow has a negative effect on a target) but I also feel that the whole mechanism should not be a safety net (which should be done before injection to me) but more a watchdog system (observing the disruption during its whole lifecycle and emitting events when something is seen).

What do you think? Should we remove it from this PR and think about a watchdog mechanism for disruptions? It would prevent to affect the reconcile loop at all because it would be totally separate.

Reply:
Im ok with this. To be honest I dont think we've ever really had an issue with a team having issues with a disruption during a sporadic environment like this, so this may have been an attempt to create a safety net for something we dont really have issues for, lots of assuming.

Its a good idea to have this in some sort of watchdog mechanism, it is great information to have as running disruptions in these environments maybe a weary operation.

I will remove Sporadic Targets in the next commit.